From b75eeeabccb2ae84d6ed798281784528772b414f Mon Sep 17 00:00:00 2001
From: "Knut St. Osmundsen" <bird@Knuts-Mac-mini.dev>
Date: Tue, 25 Jan 2022 00:42:42 +0100
Subject: [PATCH 4/4] windows build hacks

---
 YASM-VERSION-GEN.bat | 30 ++++++++++++++++++++++++++----
 1 file changed, 26 insertions(+), 4 deletions(-)

diff --git a/YASM-VERSION-GEN.bat b/YASM-VERSION-GEN.bat
index 92bb97e6..46e4861b 100644
--- a/YASM-VERSION-GEN.bat
+++ b/YASM-VERSION-GEN.bat
@@ -1,19 +1,37 @@
 @echo off
+setlocal ENABLEEXTENSIONS
+setlocal
+
+rem switch to script directory
+set MY_DIR=%~dp0
+cd %MY_DIR%
+
 if exist version goto haveversion
 
 set errorlevel=0
 set _my_=
+
 for /f "usebackq tokens=1*" %%f in (`reg query HKCU\Software\TortoiseGit /v MSysGit`) do (set _my_=%%f %%g)
-if %errorlevel% neq 0 goto notfound
-if "%_my_%" == "" goto notfound
+if %errorlevel% neq 0 goto notfound1
+if "%_my_%" == "" goto notfound1
+goto ok
+
+rem hack for my rig.
+:notfound1
+set _gitbin_=C:\Program Files\Git\usr\bin
+goto :got_git_bin
 
 rem Using the shell script version (calling Git) ...
+:ok
 set _gitbin_=%_my_:*REG_SZ=%
 for /f "tokens=* delims= " %%a in ("%_gitbin_%") do set _gitbin_=%%a
+:got_git_bin
 set OLDPATH=%PATH%
 set PATH=%_gitbin_%;%PATH%
-"%_gitbin_%\sh" YASM-VERSION-GEN.sh "%_gitbin_%"
+"%_gitbin_%\sh" -x YASM-VERSION-GEN.sh "%_gitbin_%"
 set PATH=%OLDPATH%
+endlocal
+endlocal
 exit /b
 
 :notfound
@@ -30,9 +48,13 @@ goto output
 :output
 set /p _oldver_=<YASM-VERSION-FILE
 set _oldver_=%_oldver_:~,-1%
-if "%_ver_%" == "%_oldver_%" exit /b
+if "%_ver_%" == "%_oldver_%" goto end_success
 echo %_ver_%
 echo %_ver_% > YASM-VERSION-FILE
 echo #define PACKAGE_STRING "yasm %_ver_%" > YASM-VERSION.h
 echo #define PACKAGE_VERSION "%_ver_%" >> YASM-VERSION.h
 
+:end_success
+endlocal
+endlocal
+exit /b 0
-- 
2.24.3 (Apple Git-128)

