diff -ru dita-ot-4.0.2-virgin/src/main/java/org/dita/dost/module/MoveMetaModule.java dita-ot-4.0.2-build/src/main/java/org/dita/dost/module/MoveMetaModule.java
--- dita-ot-4.0.2-virgin/src/main/java/org/dita/dost/module/MoveMetaModule.java	2023-02-18 15:43:58.000000000 +0100
+++ dita-ot-4.0.2-build/src/main/java/org/dita/dost/module/MoveMetaModule.java	2023-05-04 01:05:17.786773600 +0200
@@ -35,6 +35,20 @@
 import static org.dita.dost.util.URLUtils.toFile;
 import static org.dita.dost.util.XMLUtils.*;
 
+/* VBox -- BEGIN */
+import org.dita.dost.store.Store;
+import org.dita.dost.log.DITAOTLogger;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.xml.sax.ContentHandler;
+import org.xml.sax.XMLFilter;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import javax.xml.transform.TransformerException;
+import static org.dita.dost.util.URLUtils.setPath;
+/* VBox -- END  */
+
 /**
  * Cascades metadata from maps to topics and then from topics to maps.
  *
@@ -121,6 +135,38 @@
         }
     }
 
+    /* VBox -- START */
+    /** 
+     * Checks whether all the topics in the map refers to the same file.
+     *  
+     * This is a scenario when producing a single HTML document out of a ditamap, 
+     * and it could get very expensive to read+write the whole single.dita document 
+     * for each topic when there are many many topics (986 topics - 3-4 min). 
+     *  
+     * @param   mapSet  The map to examine.
+     * @returns Name of the target file URI if it's the same for all topics, null if 
+     *        it differs.
+     */
+    private URI getSameTopicTargetFileName(final Map<URI, Map<String, Element>> mapSet) {
+        int numTopics = 0;
+        URI prevTargetFileName = null;
+        for (final Entry<URI, Map<String, Element>> entry : mapSet.entrySet()) {
+            final URI key = stripFragment(entry.getKey());
+            final FileInfo fi = job.getFileInfo(key);
+            if (fi != null && (fi.format == null || fi.format.equals(ATTR_FORMAT_VALUE_DITA))) {
+                final URI targetFileName = job.tempDirURI.resolve(fi.uri);
+                if (prevTargetFileName == null) {
+                    prevTargetFileName = targetFileName;
+                } else if (!targetFileName.equals(prevTargetFileName)) {
+                    return null;
+                }
+                numTopics += 1;
+            }
+        }
+        return numTopics > 1 ? prevTargetFileName : null;
+    }
+    /* VBox -- END */
+
     /**
      * Push information from topicmeta in the map into the corresponding topics and maps.
      */
@@ -153,6 +199,24 @@
                 }
             }
             //process topic
+            /* VBox -- START
+               To speed up production of the single HTML where all topic entries
+               in the map refers to the same potentially gigantic dita file, we temporarily
+               replace the store to one with taylored caching for this scenario. */
+            long msElapsed = System.currentTimeMillis();
+            final Store oldStore = job.getStore();
+            CacheStoreSingleDitaFile tmpStore = null;
+            final URI sameTopicTargetFileName = getSameTopicTargetFileName(mapSet);
+            if (sameTopicTargetFileName != null) {
+                tmpStore = new CacheStoreSingleDitaFile(oldStore, sameTopicTargetFileName, logger);
+                job.setStore(tmpStore);
+                //try {
+                //    oldStore.copy(sameTopicTargetFileName,
+                //                  setPath(sameTopicTargetFileName, sameTopicTargetFileName.getPath() + "-before"));
+                //} catch (IOException oIgn) {
+                //}
+            }
+            /* VBox -- END */
             final DitaMetaWriter topicInserter = new DitaMetaWriter();
             topicInserter.setLogger(logger);
             topicInserter.setJob(job);
@@ -176,6 +240,26 @@
                     }
                 }
             }
+
+            /* VBox -- BEGIN */
+            if (tmpStore != null) {
+                try {
+                    tmpStore.flushSingleDoc();
+                } catch (IOException oXctp) {
+                    logger.error("Failed to flush " + sameTopicTargetFileName + ": " + oXctp.getMessage(), oXctp);
+                }
+            }
+            job.setStore(oldStore);
+            msElapsed = System.currentTimeMillis() - msElapsed;
+            logger.info("pushMetadata: topics - msElapsed=" + msElapsed);
+            //if (sameTopicTargetFileName != null) {
+            //    try {
+            //        oldStore.copy(sameTopicTargetFileName,
+            //                      setPath(sameTopicTargetFileName, sameTopicTargetFileName.getPath() + "-after"));
+            //    } catch (IOException oIgn) {
+            //    }
+            //}
+            /* VBox -- END */
         }
     }
 
@@ -197,4 +281,250 @@
         }
         return metaReader.getMapping();
     }
+
+    /* VBox -- BEGIN */
+    /**
+     * Caches the single HTML document during the pushMetadata topic phase. 
+     * Expecting only getDocument and writeDocument calls.
+     */
+    private final class CacheStoreSingleDitaFile implements Store {
+
+        private final Store fallbackStore;
+        private final URI singleDitaFile;
+        private Document singleDitaDoc;
+        private long lastModified;
+        private final DITAOTLogger logger;
+
+        CacheStoreSingleDitaFile(final Store fallbackStore, final URI singleDitaFile, final DITAOTLogger logger) {
+            this.fallbackStore = fallbackStore;
+            this.singleDitaFile = singleDitaFile.normalize();
+            this.singleDitaDoc = null;
+            this.lastModified = -1;
+            this.logger = logger;
+            logger.info("CacheStoreSingleDitaFile: singleDitaFile=" + this.singleDitaFile.toString());
+        }
+
+        public void flushSingleDoc() throws IOException {
+            if (singleDitaDoc != null) {
+                logger.info("CacheStoreSingleDitaFile: flush");
+                fallbackStore.writeDocument(singleDitaDoc, singleDitaFile);
+                singleDitaDoc = null;
+            }
+        }
+
+        private void flushSingleDocThrowIoXcpt(final URI path) throws IOException {
+            if (singleDitaFile.equals(path.normalize())) {
+                flushSingleDoc();
+            }
+        }
+
+        private void flushSingleDocThrowDitaXcpt(final URI path) throws DITAOTException {
+            try {
+                flushSingleDocThrowIoXcpt(path);
+            } catch (IOException oXcpt) {
+                throw new DITAOTException(oXcpt.toString());
+            }
+        }
+
+        public void transform(final URI src, final ContentHandler contentHandler) throws DITAOTException {
+            logger.error("CacheStoreSingleDitaFile: transform #1: " + src.toString());
+            assert false;
+            flushSingleDocThrowDitaXcpt(src);
+            fallbackStore.transform(src, contentHandler);
+        }
+
+        public void transform(final URI src, final List<XMLFilter> filters) throws DITAOTException {
+            logger.error("CacheStoreSingleDitaFile: transform #2: " + src.toString());
+            assert false;
+            flushSingleDocThrowDitaXcpt(src);
+            fallbackStore.transform(src, filters);
+        }
+
+        public void transform(final URI src, final URI dst, final List<XMLFilter> filters) throws DITAOTException {
+            logger.error("CacheStoreSingleDitaFile: transform #3: " + src.toString() + " => " + dst.toString());
+            assert false;
+            flushSingleDocThrowDitaXcpt(src);
+            flushSingleDocThrowDitaXcpt(dst);
+            fallbackStore.transform(src, dst, filters);
+        }
+
+        public void transform(final URI src, final XsltTransformer transformer) throws DITAOTException {
+            logger.error("CacheStoreSingleDitaFile: transform #4: " + src.toString());
+            assert false;
+            flushSingleDocThrowDitaXcpt(src);
+            fallbackStore.transform(src, transformer);
+        }
+
+        public void transform(final URI src, final URI dst, final XsltTransformer transformer) throws DITAOTException {
+            logger.error("CacheStoreSingleDitaFile: transform #5: " + src.toString() + " => " + dst.toString());
+            assert false;
+            flushSingleDocThrowDitaXcpt(src);
+            flushSingleDocThrowDitaXcpt(dst);
+            fallbackStore.transform(src, transformer);
+        }
+
+        public Source getSource(URI path) {
+            logger.error("CacheStoreSingleDitaFile: getSource " + path.toString());
+            assert false;
+            try {
+                flushSingleDocThrowIoXcpt(path);
+            } catch (IOException oXctp) {
+                logger.error("CacheStoreSingleDitaFile: flushSingleDocThrowIoXcpt failed " + oXctp.toString());
+            }
+            return fallbackStore.getSource(path);
+        }
+
+        public Document getImmutableDocument(URI path) throws IOException {
+            logger.error("CacheStoreSingleDitaFile: getImmutableDocument: " + path.toString());
+            assert false;
+            flushSingleDocThrowIoXcpt(path);
+            return fallbackStore.getImmutableDocument(path);
+        }
+
+        public XdmNode getImmutableNode(final URI path) throws IOException {
+            logger.error("CacheStoreSingleDitaFile: getImmutableNode: " + path.toString());
+            assert false;
+            flushSingleDocThrowIoXcpt(path);
+            return fallbackStore.getImmutableNode(path);
+        }
+
+        public Document getDocument(URI path) throws IOException {
+            if (singleDitaFile.equals(path.normalize())) {
+                if (singleDitaDoc != null) {
+                    logger.info("CacheStoreSingleDitaFile: getDocument hit");
+                    return singleDitaDoc;
+                }
+                logger.info("CacheStoreSingleDitaFile: getDocument first");
+                singleDitaDoc = fallbackStore.getDocument(path);
+                return singleDitaDoc;
+            }
+            logger.warn("CacheStoreSingleDitaFile: getDocument miss: " + path.toString());
+            assert false;
+            return fallbackStore.getDocument(path);
+        }
+
+        public void writeDocument(Document doc, URI dst) throws IOException {
+            if (singleDitaFile.equals(dst.normalize())) {
+                singleDitaDoc = doc;
+                lastModified = System.currentTimeMillis();
+                logger.info("CacheStoreSingleDitaFile: writeDocument hit");
+            } else {
+                logger.warn("CacheStoreSingleDitaFile: writeDocument miss: " + dst.toString());
+                fallbackStore.writeDocument(doc, dst);
+                assert false;
+            }
+        }
+
+        public void writeDocument(Node doc, final ContentHandler dst) throws IOException {
+            logger.error("CacheStoreSingleDitaFile: writeDocument miss/#1");
+            fallbackStore.writeDocument(doc, dst);
+            assert false;
+        }
+
+        public void writeDocument(XdmNode node, final ContentHandler dst) throws IOException {
+            logger.error("CacheStoreSingleDitaFile: writeDocument miss/#2");
+            fallbackStore.writeDocument(node, dst);
+            assert false;
+        }
+
+        public void writeDocument(XdmNode node, URI dst) throws IOException {
+            logger.error("CacheStoreSingleDitaFile: writeDocument miss/#3: " + dst.toString());
+            fallbackStore.writeDocument(node, dst);
+            assert false;
+        }
+
+        public Destination getDestination(URI path) throws IOException {
+            logger.error("CacheStoreSingleDitaFile: getDestination " + path.toString());
+            assert false;
+            return fallbackStore.getDestination(path);
+        }
+
+        public ContentHandler getContentHandler(URI path) throws SaxonApiException, IOException {
+            logger.error("CacheStoreSingleDitaFile: getContentHandler " + path.toString());
+            assert false;
+            flushSingleDocThrowIoXcpt(path);
+            return fallbackStore.getContentHandler(path);
+        }
+
+        public URI getUri(URI path) {
+            logger.error("CacheStoreSingleDitaFile: getUri " + path.toString());
+            assert false;
+            return fallbackStore.getUri(path);
+        }
+
+        public void delete(URI file) throws IOException {
+            logger.error("CacheStoreSingleDitaFile: delete " + file.toString());
+            assert false;
+            if (singleDitaFile.equals(file.normalize())) {
+                singleDitaDoc = null;
+                logger.info("CacheStoreSingleDitaFile: delete hit");
+            }
+            fallbackStore.delete(file);
+        }
+
+        public void copy(URI src, URI dst) throws IOException {
+            logger.error("CacheStoreSingleDitaFile: copy " + src.toString() + " -> " + dst.toString());
+            assert false;
+            flushSingleDocThrowIoXcpt(src);
+            fallbackStore.copy(src, dst);
+            if (singleDitaDoc != null && singleDitaFile.equals(dst.normalize())) {
+                singleDitaDoc = null;
+                logger.info("CacheStoreSingleDitaFile: copy hit");
+            }
+        }
+
+        public boolean exists(URI path) {
+            if (singleDitaDoc != null && singleDitaFile.equals(path.normalize())) {
+                logger.info("CacheStoreSingleDitaFile: exists hit");
+                return true;
+            }
+            logger.error("CacheStoreSingleDitaFile: exists " + path.toString());
+            assert false;
+            return fallbackStore.exists(path);
+        }
+
+        public long getLastModified(URI path) {
+            if (singleDitaDoc != null && singleDitaFile.equals(path.normalize())) {
+                logger.info("CacheStoreSingleDitaFile: getLastModified hit");
+                return lastModified;
+            }
+            logger.error("CacheStoreSingleDitaFile: getLastModified " + path.toString());
+            assert false;
+            return fallbackStore.getLastModified(path);
+        }
+
+        public void move(URI src, URI dst) throws IOException {
+            logger.error("CacheStoreSingleDitaFile: move " + src.toString() + " -> " + dst.toString());
+            assert false;
+            flushSingleDocThrowIoXcpt(src);
+            fallbackStore.move(src, dst);
+            if (singleDitaDoc != null && singleDitaFile.equals(dst.normalize())) {
+                singleDitaDoc = null;
+                logger.info("CacheStoreSingleDitaFile: move hit");
+            }
+        }
+
+        public InputStream getInputStream(URI path) throws IOException {
+            logger.error("CacheStoreSingleDitaFile: getInputStream " + path.toString());
+            assert false;
+            flushSingleDocThrowIoXcpt(path);
+            return fallbackStore.getInputStream(path);
+        }
+
+        public OutputStream getOutputStream(URI path) throws IOException  {
+            logger.error("CacheStoreSingleDitaFile: getOutputStream " + path.toString());
+            assert false;
+            flushSingleDocThrowIoXcpt(path);
+            return fallbackStore.getOutputStream(path);
+        }
+
+        /* URIResolver */
+        public Source resolve(String href, String base) throws TransformerException {
+            logger.error("CacheStoreSingleDitaFile: resolve href:" + href + " base:" + base);
+            assert false;
+            return fallbackStore.resolve(href, base);
+        }
+
+    }
+    /* VBox -- END */
 }
diff -ru dita-ot-4.0.2-virgin/src/main/java/org/dita/dost/util/Job.java dita-ot-4.0.2-build/src/main/java/org/dita/dost/util/Job.java
--- dita-ot-4.0.2-virgin/src/main/java/org/dita/dost/util/Job.java	2023-02-18 15:43:58.000000000 +0100
+++ dita-ot-4.0.2-build/src/main/java/org/dita/dost/util/Job.java	2023-05-04 01:09:18.639098100 +0200
@@ -126,7 +126,7 @@
     private final File jobFile;
     private final Map<URI, FileInfo> files = new ConcurrentHashMap<>();
     private long lastModified;
-    private final Store store;
+    private /* VBox: final*/ Store store;
 
     /**
      * Create new job configuration instance. Initialise by reading temporary configuration files.
@@ -167,6 +167,21 @@
         return store;
     }
 
+    /* VBox -- BEGIN */
+    /** 
+     * Allow changing the store.
+     * @param newStore The new store instance.
+     * @return Old store.
+     */
+    public Store setStore(Store newStore) {
+        if (newStore == null)
+            throw new IllegalArgumentException("newStore must not be null!");
+        Store oldStore = store;
+        store = newStore;
+        return oldStore;
+    }
+    /* VBox -- END */
+
     /**
      * Test if serialized configuration file has been updated.
      * @return {@code true} if configuration file has been update after this object has been created or serialized
