From c281fd3c55a6ce41950fde2a0726e7553d6af226 Mon Sep 17 00:00:00 2001
From: "knut st. osmundsen" <bird-skylake@anduin.net>
Date: Sun, 8 Mar 2020 03:07:46 +0100
Subject: [PATCH] Make_mvc.mak: Make it somewhat useful again.

---
 gsoap/src/Make_mvc.mak | 24 +++++++++++-------------
 1 file changed, 11 insertions(+), 13 deletions(-)

diff --git a/gsoap/src/Make_mvc.mak b/gsoap/src/Make_mvc.mak
index 3872f90..226d659 100644
--- a/gsoap/src/Make_mvc.mak
+++ b/gsoap/src/Make_mvc.mak
@@ -2,14 +2,14 @@ CC=cl
 CPP=cl
 LIBS=
 
-COFLAGS=-O2
+COFLAGS=-O2 -Zi -GR- -GF
 # COFLAGS=-ZI -DDEBUG
 
 CFLAGS= -EHsc -nologo $(COFLAGS)
 
-soapcpp2.exe:	soapcpp2.h soapcpp2_yacc.obj symbol2.obj error2.obj lex.yy.obj init2.obj soapcpp2.obj
-		$(CC) $(CFLAGS) symbol2.obj error2.obj soapcpp2_yacc.obj lex.yy.obj \
-			init2.obj soapcpp2.obj $(LIBS) -Fe$@
+soapcpp2.exe:	soapcpp2.h soapcpp2_yacc.tab.obj symbol2.obj error2.obj lex.yy.obj init2.obj soapcpp2.obj
+		$(CC) $(CFLAGS) symbol2.obj error2.obj soapcpp2_yacc.tab.obj lex.yy.obj \
+			init2.obj soapcpp2.obj $(LIBS) -Fe$@ -Fm -Fd -link /OPT:REF /OPT:ICF /OPT:LBR
 .c.o:		# soapcpp2.h soapcpp2_yacc.h error2.h
 		$(CC) $(CFLAGS) -DWITH_BISON -DWITH_FLEX -c $<
 clean:
@@ -17,13 +17,11 @@ clean:
 distclean:
 		del /f soapcpp2.exe *.obj *.idb *.pdb *.ilk
 real-distclean:
-		del /f soapcpp2.exe *.obj *.idb *.pdb *.ilk lex.yy.* soapcpp2_yacc.h \
-			soapcpp2_yacc.c y.tab.* *.output
+		del /f soapcpp2.exe *.obj *.idb *.pdb *.ilk lex.yy.* soapcpp2_yacc.tab.h \
+			soapcpp2_yacc.tab.c y.tab.* *.output
 
-soapcpp2_yacc.c:
-		soapcpp2_yacc.y soapcpp2.h error2.h
-		bison.exe /d /v soapcpp2_yacc.y
-		ren soapcpp2_yacc.tab.h soapcpp2_yacc.h
-		ren soapcpp2_yacc.tab.c soapcpp2_yacc.c
-lex.yy.c:	soapcpp2_yacc.h soapcpp2_yacc.c soapcpp2_lex.l
-		flex.exe /l soapcpp2_lex.l
+soapcpp2_yacc.c: soapcpp2_yacc.y soapcpp2.h error2.h
+		bison.exe -d -v soapcpp2_yacc.y
+
+lex.yy.c:	soapcpp2_yacc.tab.h soapcpp2_yacc.c soapcpp2_lex.l
+		flex.exe -l soapcpp2_lex.l
-- 
2.25.1.windows.1

