From 6b9559d9e8e1fda57dfe882a834bdd1dedf5d13f Mon Sep 17 00:00:00 2001
From: "knut st. osmundsen" <bird-skylake@anduin.net>
Date: Sun, 8 Mar 2020 19:57:02 +0100
Subject: [PATCH] stdsoap2.cpp: solaris fix

---
 gsoap/stdsoap2.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gsoap/stdsoap2.cpp b/gsoap/stdsoap2.cpp
index f8733f4..bb5f9db 100644
--- a/gsoap/stdsoap2.cpp
+++ b/gsoap/stdsoap2.cpp
@@ -5101,7 +5101,7 @@ tcp_gethostbyname(struct soap *soap, const char *addr, struct hostent *hostent,
     hostent = NULL;
     soap->errnum = h_errno;
   }
-#elif (!defined(__GLIBC__) || !defined(_GNU_SOURCE) || (!_GNU_SOURCE && !defined(_POSIX_C_SOURCE) && !defined(_XOPEN_SOURCE)) || _POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600 || defined(__ANDROID__) || defined(FREEBSD) || defined(__FreeBSD__)) && defined(HAVE_GETHOSTBYNAME_R)
+#elif (!defined(__GLIBC__) || !defined(_GNU_SOURCE) || (!_GNU_SOURCE && !defined(_POSIX_C_SOURCE) && !defined(_XOPEN_SOURCE)) || _POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600 || defined(__ANDROID__) || defined(FREEBSD) || defined(__FreeBSD__)) && defined(HAVE_GETHOSTBYNAME_R) && !defined(__sun__)
   while ((r = gethostbyname_r(addr, hostent, tmpbuf, tmplen, &hostent, &soap->errnum)) < 0)
   {
     if (tmpbuf != soap->tmpbuf)
-- 
2.25.1.windows.1

