diff -rwU1 VBoxMouse.org/abs.asm VBoxMouse/abs.asm
--- VBoxMouse.org/abs.asm	2000-09-06 11:00:24.000000000 +0200
+++ VBoxMouse/abs.asm	2026-05-06 13:23:29.110345000 +0200
@@ -57,2 +57,8 @@
    include mouse.inc
+
+IFDEF VBOXMOUSE
+   include infoseg.inc
+   include emiidc.inc                                            ;emi
+ENDIF
+
 .list
@@ -81,2 +87,6 @@
 
+IFDEF VBOXMOUSE
+       extrn  EmiSSI               : byte                             ;emi2
+       extrn  GDT_Seg         : word
+ENDIF
 
@@ -90,2 +100,3 @@
        public  Process_Absolute
+       public  _Process_Absolute        ; vboxmouse
        public  MapXY2RowCol
@@ -156,2 +167,3 @@
 
+_Process_Absolute:                      ; vboxmouse
 Process_Absolute  proc  near
@@ -273,2 +285,16 @@
 
+IFDEF VBOXMOUSE
+     push ds                    ; setup ds to prevent trap in OS/2 fullscreen
+     mov  ax, seg GDT_Seg
+     mov  ds,ax
+
+     push es                    ; setup time stamp
+     mov  es,word ptr GDT_Seg
+     mov  eax, es:SIS_MsCount
+     mov  MEvent.Mon_time,eax
+     pop  es
+
+     pop  ds
+ENDIF
+
        push NewRow
@@ -538,3 +564,2 @@
 
-       int 3
        mov  ax, RowPos
diff -rwU1 VBoxMouse.org/alltypes.c VBoxMouse/alltypes.c
--- VBoxMouse.org/alltypes.c	2000-09-06 10:54:58.000000000 +0200
+++ VBoxMouse/alltypes.c	2026-05-06 13:26:14.664881800 +0200
@@ -46,2 +46,13 @@
 #include "pdi.h"
+#ifndef NO_VA_LIST
+typedef char far *MYFARSTRPTR;
+typedef int SIZEASSERT[sizeof(MYFARSTRPTR) == 4];
+typedef int SIZEASSERT2[sizeof(__va_list) == 4];
+# ifndef __SW_ZU
+#  error Expected __SW_ZU to be defined...
+# endif
+#endif
+#ifdef VBOXMOUSE
+# include "vboxmouse.h"
+#endif
 
@@ -60,2 +71,3 @@
         "push   ds"                     \
+        "push   esi"                    \
         "mov    esi, Security_Off"      \
@@ -63,2 +75,3 @@
         "Inc    Dword Ptr [esi]"        \
+        "pop    esi"                    \
         "pop    ds";
@@ -77,2 +90,6 @@
     {
+#ifdef VBOXMOUSE
+        if (VBoxMouseSendPacketAbsolute(x, y, event))
+            return;
+#endif
         DevStatus &= ~gREADENABLE;
@@ -88,2 +105,3 @@
 
+#ifndef VBOXMOUSE
 #ifdef DEBUG
@@ -249 +267,2 @@
 #endif
+#endif /* !VBOXMOUSE */
diff -rwU1 VBoxMouse.org/alltypes.h VBoxMouse/alltypes.h
--- VBoxMouse.org/alltypes.h	2000-09-06 10:45:32.000000000 +0200
+++ VBoxMouse/alltypes.h	2026-05-06 13:26:43.438343900 +0200
@@ -14,2 +14,3 @@
 void SendPacket(SHORT x,SHORT y,UCHAR event);
+#ifndef VBOXMOUSE
 void _cdecl sprf(char _far *buf,char _far *stg,char _far *parms);
@@ -20,2 +21,3 @@
 #endif
+#endif
 
diff -rwU1 VBoxMouse.org/cmdline.c VBoxMouse/cmdline.c
--- VBoxMouse.org/cmdline.c	2000-09-06 10:55:16.000000000 +0200
+++ VBoxMouse/cmdline.c	2020-03-04 19:29:25.625703300 +0100
@@ -46,2 +46,3 @@
 #include "pdi.h"
+#include "VBoxMouse.h"
 
@@ -88,3 +89,3 @@
 **   EFFECTS - gNumTabs - global variable, containing number of arrows 
-**             being injected on mouse`s stick <TAB>/<SHIFT+TAB> events
+**             being injected on mouse`s stick \<TAB\>/\<SHIFT+TAB\> events
 **   
@@ -137,5 +138,3 @@
 
-	#ifdef DEBUG
-	prf("MOUSE : CheckForTabs : found=%d gNumTabs=%d\r\n", bFound, gNumTabs);
-	#endif
+	dprintf(("MOUSE : CheckForTabs : found=%d gNumTabs=%d\n", bFound, gNumTabs));
 
@@ -154,5 +153,4 @@
 
-#ifdef DEBUG
-	prf("Mouse CONFIG.SYS line: '%s'\r\n",(char __far *)stg);
-#endif
+	dprintf(("Mouse CONFIG.SYS line: '%s'\n", (char __far *)stg));
+
 	while ( *stg != '\0' )
@@ -182,5 +180,3 @@
 	}
-#ifdef DEBUG
-	prf("   argc=%x\r\n",argc);
-#endif
+	dprintf(("   argc=%d\n", argc));
 
@@ -188,6 +184,3 @@
 	{
-
-#ifdef DEBUG
-		prf("   argv[%x]=%s\r\n",i,(char __far *)argv[i]);
-#endif
+		dprintf(("   argv[%d]=%s\n", i, argv[i]));
 
@@ -197,5 +190,3 @@
 			gucFindMouse=FALSE;
-			#ifdef DEBUG
-			prf("MOUSE.SYS : FINDMOUSE OFF\r\n");
-			#endif
+			dprintf(("MOUSE.SYS: FINDMOUSE OFF\n"));
 		}
@@ -209,5 +200,3 @@
 
-			#ifdef DEBUG
-			prf("MOUSE.SYS : NUMTABS=%d\r\n", gNumTabs);
-			#endif
+			dprintf(("MOUSE.SYS: NUMTABS=%d\n", gNumTabs));
 		}
@@ -257,3 +246,10 @@
 		}
+#ifdef VBOXMOUSE
+		else if ( !_fstrnicmp(argv[i],"SLOWINIT",8) )
+		{
+			fVBoxSlowInit=TRUE;
+		}
+#endif
 #ifdef DEBUG
+# ifndef VBOXMOUSE
 		else if ( !_fstrnicmp(argv[i],"DEBUG=",4) )
@@ -263,7 +259,8 @@
 			debug_comport=gethex(&argv[i][6]);
-			prf("*****\r\n***** MOUSE.SYS Debugging Enabled\r\n*****\r\n");
+			dprintf(("*****\n***** MOUSE.SYS Debugging Enabled\n*****\n"));
 		}
+# endif
 		else if ( !_fstrnicmp(argv[i],"INT3",4) )
 		{
-			prf("***** CONFIG.SYS TRIGGERED INT3 IN MOUSE.SYS INIT CODE! *****\r\n");
+			dprintf(("***** CONFIG.SYS TRIGGERED INT3 IN MOUSE.SYS INIT CODE! *****\n"));
 			INT3();
Only in VBoxMouse/: crtstuff.c
diff -rwU1 VBoxMouse.org/data.asm VBoxMouse/data.asm
--- VBoxMouse.org/data.asm	2000-09-06 11:00:38.000000000 +0200
+++ VBoxMouse/data.asm	2026-05-06 13:27:42.194022700 +0200
@@ -371,2 +371,5 @@
 
+ifdef VBOXMOUSE
+_fVBoxSlowInit   db     FALSE         ; Skip all the tedious slow crap like serial
+endif
 
@@ -681,2 +684,5 @@
        public  fNullMouseParm                                              ;SRD
+ifdef VBOXMOUSE
+       public _fVBoxSlowInit
+endif
 
diff -rwU1 VBoxMouse.org/devhelp.h VBoxMouse/devhelp.h
--- VBoxMouse.org/devhelp.h	2026-05-06 14:02:01.280043200 +0200
+++ VBoxMouse/devhelp.h	2020-03-04 19:29:25.683761400 +0100
@@ -38,4 +38,19 @@
 
+#if 1
+# define DEVHELP_JUST_CALL \
+   "call dword ptr [Device_Help]"
+# define DEVHELP_JUST_CALL_ES \
+   "call dword ptr es:[Device_Help]"
+# define DEVHELP_JUST_JMP_ES \
+   "jmp dword ptr es:[Device_Help]"
+#else
+# define DEVHELP_JUST_CALL \
+   "call far [Device_Help]"
+# define DEVHELP_JUST_CALL_ES \
+   "call far es:[Device_Help]"
+# define DEVHELP_JUST_JMP_ES \
+   "jmp far es:[Device_Help]"
+#endif
 #define DEVHELP_CALL \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -65,3 +80,3 @@
    "mov  dl,63h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "pop  bx" \
@@ -89,3 +104,3 @@
    "mov  dl,18h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -105,3 +120,3 @@
    "mov  dl,0Dh", \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -140,3 +155,3 @@
    "mov  dl,2Ah" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "mov  ax,0" \
@@ -166,3 +181,3 @@
    "mov  dl,21h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -192,3 +207,3 @@
    "xchg ax,bx" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -204,3 +219,3 @@
    "mov  dl,31h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "sub  ax,ax" \
@@ -237,3 +252,3 @@
    "mov  dl,19h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "mov  ax,0" \
@@ -247,3 +262,3 @@
    "mov  dl,0Eh", \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "sub  ax,ax" \
@@ -271,3 +286,3 @@
    "mov  dl,5Dh" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -285,3 +300,3 @@
    "mov  dl,38h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -310,3 +325,3 @@
    "mov  dl,24h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "mov  es:[di],bx" \
@@ -321,3 +336,3 @@
    "mov  dl,34h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -337,3 +352,3 @@
    "mov  dl,2Bh" \
-   "jmp far es:[Device_Help]" \
+   DEVHELP_JUST_JMP_ES \
    parm [es si] [di] \
@@ -347,3 +362,3 @@
    "mov  ebx,ss:[bx]" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -363,3 +378,3 @@
    "mov  dl,13h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -383,3 +398,3 @@
    "mov  dl,1Fh" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -432,3 +447,3 @@
    "mov  dl,60h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -445,3 +460,3 @@
    "mov  dl,5Fh" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -469,3 +484,3 @@
    "mov  dh,[bp+12]" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -497,3 +512,3 @@
    "mov  dl,17h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -516,3 +531,3 @@
    "push ds" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -546,3 +561,3 @@
    "xchg cx,di" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -558,3 +573,3 @@
    "xchg ax,bx" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "mov  es:[si],ax" \
@@ -576,3 +591,3 @@
    "mov  dl,0Ah" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -594,3 +609,3 @@
    "mov  dl,09h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "sub  ax,ax" \
@@ -603,3 +618,3 @@
    "mov  dl,10h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "sub  ax,ax" \
@@ -619,3 +634,3 @@
    "mov  dl,0Fh" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "sub  ax,ax" \
@@ -628,3 +643,3 @@
    "mov  dl,12h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -651,3 +666,3 @@
    "mov  dl,28h" \
-   "call far es:[Device_Help]" \
+   DEVHELP_JUST_CALL_ES \
    "push es" \
@@ -673,3 +688,3 @@
    "mov  dl,51h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "sub  ax,ax" \
@@ -686,3 +701,3 @@
    "xchg ax,bx" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -716,3 +731,3 @@
    "mov  dl,3Ah" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "mov  ax,0" \
@@ -726,3 +741,3 @@
    "mov  dl,61h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "mov  ax,bx" \
@@ -772,3 +787,3 @@
    "mov  dl,0h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "sub  ax,ax" \
@@ -793,3 +808,3 @@
    "mov  dl,8h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -824,3 +839,3 @@
    "mov  dl,25h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "mov  ax,0" \
@@ -848,3 +863,3 @@
    "mov  dl,0Ch" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "sub  ax,ax" \
@@ -857,3 +872,3 @@
    "mov  dl,3" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "sub  ax,ax" \
@@ -901,3 +916,3 @@
    "mov  dl,3Ch" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "mov  ax,0" \
@@ -913,3 +928,3 @@
    "mov  dl,5bh" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -932,3 +947,3 @@
    "mov  si,ax" \
-   "call far es:[Device_Help]" \
+   DEVHELP_JUST_CALL_ES \
    "pop  es" \
@@ -959,3 +974,3 @@
    "mov  dl,57h" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -991,3 +1006,3 @@
    "mov  ecx,[esp+8]" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -1016,3 +1031,3 @@
    "mov  esi,[esp+16]" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -1035,3 +1050,3 @@
    "mov  ecx,[esp+8]" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "jc   error" \
@@ -1070,3 +1085,3 @@
    "mov  dl,2" \
-   "call far [Device_Help]" \
+   DEVHELP_JUST_CALL \
    "sub  ax,ax" \
diff -rwU1 VBoxMouse.org/kbd.c VBoxMouse/kbd.c
--- VBoxMouse.org/kbd.c	2000-09-06 10:55:26.000000000 +0200
+++ VBoxMouse/kbd.c	2026-05-06 13:30:30.078260300 +0200
@@ -108,2 +108,7 @@
 
+#ifdef VBOXMOUSE /* bird: the scrolling causes a weird (e)bp corruption here.
+                  * rewrote it in assembly (util1.asm) to avoid any isses with
+                  * inline assmbly / watcom. */
+extern USHORT __cdecl Call_KbdDI(int func, char __far *buffer);
+#else /* !VBOXMOUSE */
 USHORT
@@ -157,2 +162,3 @@
 }
+#endif /* !VBOXMOUSE */
 
@@ -164,3 +170,3 @@
 
-	if ( !kbd_flags & ATTACHED )
+    if ( !(kbd_flags & ATTACHED) ) /* vboxmouse: missing parentheses */
 	{
@@ -186,2 +192,3 @@
 {
+#if 0
 #pragma aux KbdAsm3 =                   \
@@ -189,2 +196,3 @@
    "mov   KbdDI_Entry, edx";
+#endif
 
@@ -192,3 +200,3 @@
 
-	if ( !kbd_flags & ATTACHED )
+    if ( !(kbd_flags & ATTACHED) )
 	{
@@ -198,2 +206,3 @@
 			kbd_flags |= ATTACHED;
+#if 0
 			/*
@@ -206,2 +215,5 @@
 			KbdAsm3();
+#else
+                        *(PULONG)&KbdDI_Entry = KbdDD.protentry;
+#endif
 		}
diff -rwU1 VBoxMouse.org/macs.inc VBoxMouse/macs.inc
--- VBoxMouse.org/macs.inc	2000-09-06 11:02:52.000000000 +0200
+++ VBoxMouse/macs.inc	2026-05-06 13:31:23.814221800 +0200
@@ -138,2 +138,6 @@
 
+ifdef VBOXMOUSE ; We can dispense with the I/O delay.
+MyIOdelay   MACRO
+            ENDM
+else
 MyIOdelay MACRO
@@ -145,2 +149,3 @@
         ENDM
+endif
 
Only in VBoxMouse.org/: makefile
Only in VBoxMouse/: Makefile.kmk
diff -rwU1 VBoxMouse.org/mou.h VBoxMouse/mou.h
--- VBoxMouse.org/mou.h	2000-09-06 10:48:30.000000000 +0200
+++ VBoxMouse/mou.h	2020-03-04 19:29:25.618696100 +0100
@@ -99,3 +99,3 @@
    "mov  ax,1"         \
-   "call far GIDC_Entry";
+   "call far ptr GIDC_Entry";
 
@@ -247,2 +247,5 @@
 extern USHORT gNumTabs;
+#ifdef VBOXMOUSE
+extern char   fVBoxSlowInit;
+#endif
 
Only in VBoxMouse.org/: mouse.def
diff -rwU1 VBoxMouse.org/pdic.c VBoxMouse/pdic.c
--- VBoxMouse.org/pdic.c	2000-09-06 10:56:04.000000000 +0200
+++ VBoxMouse/pdic.c	2026-05-06 13:34:08.052912500 +0200
@@ -50,2 +50,3 @@
 #include "pdiproto.h"
+#include "VBoxMouse.h"
 
@@ -58,3 +59,19 @@
 extern UCHAR gucFindMouse;
+#ifndef VBOX
 #define DEBUG_POST(val) outp(0x80,val);
+#else
+static void debugPost(unsigned iCode)
+{
+    static char const s_szHex[] = "0123456789abcdef";
+    outp(0x504, 'M');
+    outp(0x504, 'O');
+    outp(0x504, 'U');
+    outp(0x504, ':');
+    outp(0x504, s_szHex[(iCode >> 4) & 0xf]);
+    outp(0x504, s_szHex[iCode & 0xf]);
+    outp(0x504, '\n');
+}
+# define DEBUG_POST(val) debugPost(val)
+# define DEBUG_BIRD(ch) do { outp(0x504, ch); outp(0x504, '\n'); } while (0)
+#endif
 
@@ -84,3 +101,3 @@
    "mov  dl,63h" \
-   "call far [Device_Help]" \
+   "call far ptr [Device_Help]" \
    "pop  bx" \
@@ -111,3 +128,3 @@
 #ifdef DEBUG_L0
-	prf("   ClearAuxPort\r\n");
+    dprintf(("   ClearAuxPort\n"));
 #endif
@@ -141,3 +158,3 @@
 #ifdef DEBUG_L0
-	prf("   Read8042Data");
+    dprintf(("   Read8042Data"));
 #endif
@@ -151,3 +168,3 @@
 #ifdef DEBUG_L0
-			prf("->%x\r\n",c);
+            dprintf(("->%x\n",c));
 #endif
@@ -158,3 +175,3 @@
 #ifdef DEBUG
-	prf("***READ8042DATA ERROR***\r\n");
+    dprintf((" - ***READ8042DATA ERROR***\n"));
 #endif
@@ -188,3 +205,3 @@
 #ifdef DEBUG_L0
-	prf("   WriteControl(%x)\r\n",byte);
+    dprintf(("   WriteControl(%x)\n",byte));
 #endif
@@ -200,3 +217,3 @@
 #ifdef DEBUG
-	prf("***WRITECONTROLREGISTER ERROR***\r\n");
+    dprintf(("***WRITECONTROLREGISTER ERROR***\n"));
 #endif
@@ -240,3 +257,3 @@
 #ifdef DEBUG_L0
-	prf("   WriteInput(%x)\r\n",byte);
+    dprintf(("   WriteInput(%x)\n",byte));
 #endif
@@ -251,3 +268,3 @@
 #ifdef DEBUG
-	prf("***WRITEINPUTBUFFER ERROR***\r\n");
+    dprintf(("***WRITEINPUTBUFFER ERROR***\n"));
 #endif
@@ -261,3 +278,3 @@
 #ifdef DEBUG_L1
-	prf("SetCommand(%x)\r\n",val);
+    dprintf(("SetCommand(%x)\n",val));
 #endif
@@ -271,3 +288,3 @@
 #ifdef DEBUG
-	prf("***SETCOMMAND ERROR***\r\n");
+    dprintf(("***SETCOMMAND ERROR***\n"));
 #endif
@@ -290,3 +307,3 @@
 #ifdef DEBUG_L1
-	prf("CheckForAux\r\n");
+    dprintf(("CheckForAux\n"));
 #endif
@@ -297,3 +314,5 @@
 	if ( Read8042Data() == -1 ) return (0);
+#ifndef VBOX
 	DEBUG_POST(5);
+#endif
 	return (1);
@@ -310,5 +329,3 @@
 	CLI();
-#ifdef DEBUG
-	prf("   EnableKbdInt\r\n");
-#endif
+    dprintf(("   EnableKbdInt\n"));
 	if ( 1  /* !smp */ )
@@ -339,5 +356,3 @@
 	CLI();
-#ifdef DEBUG
-	prf("   DisableKbdInt\r\n");
-#endif
+    dprintf(("   DisableKbdInt\n"));
 	if ( 1  /* !smp */ )
@@ -387,3 +402,3 @@
 #ifdef DEBUG_L0
-	prf("   GetMouseData->");
+    dprintf(("   GetMouseData->"));
 #endif
@@ -397,4 +412,4 @@
 #ifdef DEBUG_L0
-				prf("%x ",c);
-				prf("  status=%x\r\n",inp(PDI_STATPORT));
+                dprintf(("%x ",c));
+                dprintf(("  status=%x\n",inp(PDI_STATPORT)));
 #endif
@@ -405,3 +420,3 @@
 #ifdef DEBUG_L0
-	prf("ERROR\r\n");
+    dprintf(("ERROR\n"));
 #endif
@@ -437,3 +452,3 @@
 #ifdef DEBUG_L0
-	prf("   GetMouseDataLONG->");
+    dprintf(("   GetMouseDataLONG->"));
 #endif
@@ -448,4 +463,4 @@
 #ifdef DEBUG_L0
-				prf("%x ",c);
-				prf("  status=%x\r\n",inp(PDI_STATPORT));
+                dprintf(("%x ",c));
+                dprintf(("  status=%x\n",inp(PDI_STATPORT)));
 #endif
@@ -456,3 +471,3 @@
 #ifdef DEBUG_L0
-	prf("ERROR\r\n");
+    dprintf(("ERROR\n"));
 #endif
@@ -471,13 +486,17 @@
 
-	prf("8042: CtlStat=%c ",inp(0x64));
+    dprintf(("8042: CtlStat=%x ",inp(0x64)));
 	WriteControlRegister(CMD_READINP);
-	prf("InpPort=%c ",Read8042Data());
+    dprintf(("InpPort=%x ",Read8042Data()));
 	WriteControlRegister(CMD_READOUT);
-	prf("OutPort=%c\r\nMemory: ",Read8042Data());
+    dprintf(("OutPort=%x\nMemory: ",Read8042Data()));
+# ifdef VBOXMOUSE /* avoid nasty errors in Read8042Data... */ /** @todo check this on real hw. */
+    for ( i=0 ; i < 0x1 ; i++ )
+# else
 	for ( i=0 ; i < 0x1f ; i++ )
+# endif
 	{
 		WriteControlRegister(CMD_READ_CMD+i);
-		prf("%c ",Read8042Data());
+        dprintf(("%x ",Read8042Data()));
 	}
-	prf("\r\n");
+    dprintf(("\n"));
 }
@@ -507,3 +526,3 @@
 #ifdef DEBUG_L1
-	prf("SendToMouse(%x)\r\n",byte);
+    dprintf(("SendToMouse(%x)\n",byte));
 #endif
@@ -531,3 +550,3 @@
 #ifdef DEBUG_L2
-	prf("EnablePDI\r\n");
+    dprintf(("EnablePDI\n"));
 #endif
@@ -538,3 +557,3 @@
 #ifdef DEBUG_L2
-		prf("%x\r\n",inp(0x60));
+        dprintf(("%x\n",inp(0x60)));
 #else
@@ -569,3 +588,3 @@
 #ifdef DEBUG_L2
-	prf("ResetMouse()\r\n");
+    dprintf(("ResetMouse()\n"));
 #endif
@@ -695,3 +714,3 @@
 #ifdef DEBUG_L1
-	prf("GetTranslate() raw->%x\r\n",ccb);
+    dprintf("GetTranslate() raw->%x\n",ccb));
 #endif
@@ -721,4 +740,3 @@
 
-#ifdef DEBUG
-	prf("APMResume\r\n");
+    dprintf(("APMResume\n"));
 #ifdef DEBUG_8042
@@ -726,3 +744,2 @@
 #endif
-#endif
 
@@ -747,5 +764,3 @@
 
-		#ifdef DEBUG
-		prf(" scale=%d resol=%d sampl=%d\r\n", scale, resol, sampl);
-		#endif
+        dprintf((" scale=%d resol=%d sampl=%d\n", scale, resol, sampl));
 
@@ -756,9 +771,5 @@
 			{
-				#ifdef DEBUG
-				prf("APMResume : Primax mouse detection completed\r\n");
-				#endif
+                dprintf(("APMResume : Primax mouse detection completed\n"));
 				CheckforPDIDevice();
-				 #ifdef DEBUG
-				prf("APMResume : CheckForPDIDevice completed\r\n");
-				#endif
+                dprintf(("APMResume : CheckForPDIDevice completed\n"));
 				break;
@@ -810,4 +821,3 @@
    
-#ifdef DEBUG
-	prf("CheckforPDIDevice-start\r\n");
+    dprintf(("CheckforPDIDevice-start\n"));
 #ifdef DEBUG_8042
@@ -815,3 +825,2 @@
 #endif
-#endif
 	/*************************************************************************
@@ -822,3 +831,3 @@
 	 * test AUX interface command (0a9h) as if they had a system mouse port
-	 * when in fact they did not. When the preceeding AUX commands were sent
+     * when in fact they did not. When the preceding AUX commands were sent
 	 * to the controller the 8042 clone chip was leaving the keyboard in a
@@ -928,4 +937,3 @@
 		EnableKbdInt();
-#ifdef DEBUG
-		prf("CheckforPDIDevice-return(1)\r\n");
+        dprintf(("CheckforPDIDevice-return(1)\n"));
 #ifdef DEBUG_8042
@@ -933,7 +941,5 @@
 #endif
-#endif
 		return (1);
 	}
-#ifdef DEBUG
-	prf("CheckforPDIDevice-return(0)\r\n");
+    dprintf(("CheckforPDIDevice-return(0)\n"));
 #ifdef DEBUG_8042
@@ -941,3 +947,2 @@
 #endif
-#endif
 	return (0);
@@ -945,5 +950,3 @@
 	error:
-#ifdef DEBUG
-	prf("CheckforPDIDevice-error\r\n");
-#endif
+    dprintf(("CheckforPDIDevice-error\n"));
 	if ( AUXPortExists )
@@ -973,4 +976,3 @@
 	pdi_exit:
-#ifdef DEBUG
-	prf("CheckforPDIDevice-pdi_exit\r\n");
+    dprintf(("CheckforPDIDevice-pdi_exit\n"));
 #ifdef DEBUG_8042
@@ -978,3 +980,2 @@
 #endif
-#endif
 	EnableKbdInt();
@@ -1017,5 +1018,3 @@
 
-#ifdef DEBUG
-	prf("PDIWaitForPW\r\n");
-#endif
+    dprintf(("PDIWaitForPW\n"));
 
@@ -1054,4 +1053,3 @@
 	PUSHA();
-#ifdef DEBUG
-	prf("HookHotPlugPDI-start\r\n");
+    dprintf(("HookHotPlugPDI-start\n"));
 #ifdef DEBUG_8042
@@ -1059,7 +1057,5 @@
 #endif
-#endif
 	CheckforPDIDevice();
 	SynchFlag &= ~HOTPLUG_MASK;
-#ifdef DEBUG
-	prf("HookHotPlugPDI-exit\r\n");
+    dprintf(("HookHotPlugPDI-exit\n"));
 #ifdef DEBUG_8042
@@ -1067,3 +1063,2 @@
 #endif
-#endif
 	POPA();
@@ -1121,5 +1116,3 @@
 	{
-#ifdef DEBUG
-		prf("*** Out-Of-Sync timeout occurred! Delta=%d\r\n", Delta);
-#endif
+        dprintf(("*** Out-Of-Sync timeout occurred! Delta=%d\n", Delta));
 		ByteCount = 0;
@@ -1140,5 +1133,3 @@
 			SynchFlag |= HOTPLUG2;
-#ifdef DEBUG
-			prf("*** Mouse HOTPLUG confirmed!\r\n");
-#endif
+            dprintf(("*** Mouse HOTPLUG confirmed!\n"));
 			if ( HotPlugHandle != 0 )
diff -rwU1 VBoxMouse.org/pdiintel.c VBoxMouse/pdiintel.c
--- VBoxMouse.org/pdiintel.c	2000-09-06 10:56:34.000000000 +0200
+++ VBoxMouse/pdiintel.c	2026-05-06 13:34:15.186110700 +0200
@@ -19,2 +19,3 @@
 #include "pdiproto.h"
+#include "VBoxMouse.h"
 
@@ -37,3 +38,3 @@
 #ifdef DEBUG_L2
-    prf("MSKnock\r\n");
+    dprintf(("MSKnock\n");
 #endif
@@ -153,3 +154,3 @@
 #ifdef DEBUG_PACKET
-            prf("Processing Wheel Packet (0,0,%x)\r\n", z);
+            dprintf(("Processing Wheel Packet (0,0,%x)\n", z);
 #endif
diff -rwU1 VBoxMouse.org/pdiproto.c VBoxMouse/pdiproto.c
--- VBoxMouse.org/pdiproto.c	2000-09-06 10:56:44.000000000 +0200
+++ VBoxMouse/pdiproto.c	2026-05-06 13:35:06.636099500 +0200
@@ -1,2 +1,2 @@
-/*SCCSID:  src/dev/mouse/scroll/pdiproto.c, mse, c.basedd  98/12/16                                                     *
+/*SCCSID:  src/dev/mouse/scroll/pdiproto.c, mse, c.basedd  98/12/16           */
 /************************** START OF SPECIFICATIONS ***************************/
@@ -67,2 +67,3 @@
 #include "alltypes.h"
+#include "VBoxMouse.h"
 
@@ -77,3 +78,3 @@
 UCHAR          gucFindMouse=TRUE;
-USHORT         type=0;subtype=0;
+USHORT         type=0, subtype=0;
 USHORT         type2=0,buttn=0,firm1=0;
@@ -92,2 +93,3 @@
 USHORT  PS2_PrimaxInit(USHORT rate, USHORT resol, USHORT scale);
+void    IODelay(void);
 #endif
@@ -187,3 +189,3 @@
 #ifdef DEBUG_L2
-        prf("ResetPrescott\r\n");
+        dprintf(("ResetPrescott\n"));
 #endif
@@ -194,3 +196,3 @@
 #ifdef DEBUG_L2
-        prf("...first byte was %x\r\n", c);
+        dprintf(("...first byte was %x\n", c));
 #endif
@@ -204,3 +206,3 @@
 #ifdef DEBUG_L2
-        prf("...second byte was %x\r\n", c);
+        dprintf(("...second byte was %x\n", c));
 #endif
@@ -215,3 +217,3 @@
 #ifdef DEBUG_L2
-                prf("...third byte was %x\r\n", c);
+                dprintf(("...third byte was %x\n", c));
 #endif
@@ -349,3 +351,3 @@
 #ifdef DEBUG_DETECT
-        prf("CollectPDIPnp(%d)\r\n", slot);
+        dprintf(("CollectPDIPnp(%d)\n", slot));
 #endif
@@ -367,3 +369,3 @@
 #ifdef DEBUG_DETECT
-        prf("Mouse Pnp String='%s'\r\n", buffer);
+        dprintf(("Mouse Pnp String='%s'\n", buffer));
 #endif
@@ -373,4 +375,4 @@
           if ( *ptr == '(' ) {
-	    man = (ptr[3] << 16) + (ptr[4] << 8) + (ptr[5]);
-	    pro = (ptr[6] << 24) + (ptr[7] << 16) + (ptr[8] << 8) + (ptr[9]);
+	    man = ((ULONG)ptr[3] << 16) + ((ULONG)ptr[4] << 8) + (ptr[5]);
+	    pro = ((ULONG)ptr[6] << 24) + ((ULONG)ptr[7] << 16) + ((ULONG)ptr[8] << 8) + (ptr[9]);
           }
@@ -378,3 +380,3 @@
 #ifdef DEBUG_DETECT
-        prf("ParsePnp says Man=%x Pro=%x\r\n", man, pro);
+        dprintf(("ParsePnp says Man=%x Pro=%x\n", man, pro));
 #endif
@@ -395,2 +397,3 @@
    USHORT PacketBuffer[256];
+#if 0 /* unused */
    USHORT Byte1=0;
@@ -398,2 +401,3 @@
    USHORT Byte3=0;
+#endif
    USHORT type=0;
@@ -456,5 +460,3 @@
   { PacketBuffer[j]=inp(0x60);
-#ifdef DEBUG
-prf("CheckForLgtchWrlss: Received %x byte...\r\n",PacketBuffer[j]);
-#endif
+dprintf(("CheckForLgtchWrlss: Received %x byte...\n",PacketBuffer[j]));
 IODelay();
@@ -469,5 +471,3 @@
    goto err; 
-#ifdef DEBUG
-prf("CheckForLgtch: Received %x bytes. \r\n",j);
-#endif
+dprintf(("CheckForLgtch: Received %x bytes. \n",j));
 
@@ -481,5 +481,3 @@
 err:
-#ifdef DEBUG
-   prf("CheckForLgtchWrlss: Error recognizing Logitech wireless scrolling mouse\r\n");
-#endif
+   dprintf(("CheckForLgtchWrlss: Error recognizing Logitech wireless scrolling mouse\n"));
    return (DETECT_ERROR);
@@ -514,5 +512,3 @@
 
-        #ifdef DEBUG
-        prf("CheckForPrimax entry\r\n");
-        #endif
+        dprintf(("CheckForPrimax entry\n"));
 
@@ -632,3 +628,4 @@
 
-        if ( usType = GetMouseData() == -1 )
+        usType = GetMouseData();
+        if ( usType == -1 )
         {
@@ -648,5 +645,3 @@
         err:
-#ifdef DEBUG
-        prf("CheckForPrimax: Error recognizing Primax scrolling mouse\r\n");
-#endif
+        dprintf(("CheckForPrimax: Error recognizing Primax scrolling mouse\n"));
         return (DETECT_ERROR);
@@ -656,5 +651,3 @@
 
-        #ifdef DEBUG
-        prf("CheckForPrimax OK exiting....\r\n");
-        #endif
+        dprintf(("CheckForPrimax OK exiting....\n"));
 
@@ -665,3 +658,3 @@
 
-CalculateType(USHORT slot)
+int CalculateType(USHORT slot)
 {
@@ -677,3 +670,3 @@
 #ifdef DEBUG_DETECT
-        prf("       Understood MOUCMD_READTYPE\r\n");
+        dprintf(("       Understood MOUCMD_READTYPE\n"));
 #endif
@@ -681,2 +674,10 @@
         if ( (type = GetMouseData()) == -1 )                     goto step2;
+#ifdef VBOXMOUSE
+        /* This will save us a 64K*10 kbd I/O operations. */
+        /** @todo It seems we are somehow vulnerable during this, I've seen (v)mouse failures
+         * and similar I touch the mouse, keyboard or switch focus (the problematic key
+         * up / reset fun stuff from the GUI) after OS/2 logo goes into text mode. */
+        if ( !fVBoxSlowInit )
+            goto step2;
+#endif
         if ( SendToMouse(MOUCMD_READID2) )                               goto step1b;
@@ -684,3 +685,3 @@
 #ifdef DEBUG_DETECT
-        prf("       Understood MOUCMD_READID2\r\n");
+        dprintf(("       Understood MOUCMD_READID2\n"));
 #endif
@@ -690,3 +691,3 @@
 #ifdef DEBUG_DETECT
-        prf("Subtype=%x\r\n", subtype);
+        dprintf(("Subtype=%x\n", subtype));
 #endif
@@ -695,3 +696,3 @@
 #ifdef DEBUG_DETECT
-        prf("step1b:\r\n");
+        dprintf(("step1b:\n"));
 #endif
@@ -699,3 +700,3 @@
 #ifdef DEBUG_DETECT
-        prf("      Not a Pnp Device\r\n");
+        dprintf(("      Not a Pnp Device\n"));
 #endif
@@ -708,3 +709,3 @@
 #ifdef DEBUG_DETECT
-                prf("       Found a Prescott\r\n");
+                dprintf(("       Found a Prescott\n"));
 #endif
@@ -720,3 +721,3 @@
 #ifdef DEBUG_DETECT
-        prf("step2:\r\n");
+        dprintf(("step2:\n"));
 #endif
@@ -728,3 +729,3 @@
 #ifdef DEBUG_DETECT
-        prf("       Understood MOUCMD_STATUSREQ\r\n");
+        dprintf(("       Understood MOUCMD_STATUSREQ\n"));
 #endif
@@ -737,4 +738,4 @@
 #ifdef DEBUG_DETECT
-                prf(" buttn=%x\r\n", buttn);
-                prf(" firm1=%x\r\n", firm1);
+                dprintf((" buttn=%x\n", buttn));
+                dprintf((" firm1=%x\n", firm1));
 #endif
@@ -745,3 +746,3 @@
 #ifdef DEBUG_DETECT
-                prf("step2b:\r\n");
+                dprintf(("step2b:\n");
 #endif
@@ -754,3 +755,3 @@
 #ifdef DEBUG_DETECT
-                prf("       Understood Enhanced MOUCMD_STATUSREQ\r\n");
+                dprintf(("       Understood Enhanced MOUCMD_STATUSREQ\n"));
 #endif
@@ -761,5 +762,5 @@
 #ifdef DEBUG_DETECT
-                prf(" ergo=%x\r\n", ergo);
-                prf(" wire=%x\r\n", wire);
-                prf(" firm2=%x\r\n", firm2);
+                dprintf((" ergo=%x\n", ergo));
+                dprintf((" wire=%x\n", wire));
+                dprintf((" firm2=%x\n", firm2));
 #endif
@@ -773,3 +774,3 @@
 #ifdef DEBUG_DETECT
-        prf("step3:\r\n");
+        dprintf(("step3:\n"));
 #endif
@@ -780,3 +781,3 @@
 #ifdef DEBUG_DETECT
-        prf("       Sample200\r\n");
+        dprintf(("       Sample200\n"));
 #endif
@@ -785,3 +786,3 @@
 #ifdef DEBUG_DETECT
-        prf("       Sample100\r\n");
+        dprintf(("       Sample100\n"));
 #endif
@@ -790,3 +791,3 @@
 #ifdef DEBUG_DETECT
-        prf("       Sample80\r\n");
+        dprintf(("       Sample80\n"));
 #endif
@@ -795,3 +796,3 @@
 #ifdef DEBUG_DETECT
-        prf("       READTYPE\r\n");
+        dprintf(("       READTYPE\n"));
 #endif
@@ -802,3 +803,3 @@
 #ifdef DEBUG_DETECT
-        prf("       Understood MOUCMD_READTYPE\r\n");
+        dprintf(("       Understood MOUCMD_READTYPE\n"));
 #endif
@@ -812,3 +813,3 @@
 #ifdef DEBUG_DETECT
-        prf("step4: RESET!\r\n");
+        dprintf(("step4: RESET!\n"));
 #endif
@@ -819,5 +820,5 @@
 #ifdef DEBUG_DETECT
-        prf(" type1=%x\r\n",type);
-        prf(" type2=%x\r\n",type2);
-        prf(" type3=%x\r\n",type3);
+        dprintf((" type1=%x\n",type));
+        dprintf((" type2=%x\n",type2));
+        dprintf((" type3=%x\n",type3));
 #endif
@@ -873,2 +874,7 @@
 
+#ifdef VBOXMOUSE /* save a little bit, nothing much compared to what we do in CalculateType(). */
+    rc = NOT_PRIMAX;
+    if (fVBoxSlowInit)
+    {
+#endif
         rc= CheckForLgtchWrlss();  // Added Logitech Wireless Scroll Mouse support
@@ -890,2 +896,5 @@
         }
+#ifdef VBOXMOUSE
+    }
+#endif
 
@@ -908,4 +917,4 @@
 #ifdef DEBUG_DETECT
-        prf(" Manufacturer[0]=%X Product[0]=%X\r\n",Manufac[0],Product[0]);
-        prf(" Manufacturer[1]=%X Product[1]=%X\r\n",Manufac[1],Product[1]);
+        dprintf((" Manufacturer[0]=%lx Product[0]=%lx\n",Manufac[0],Product[0]));
+        dprintf((" Manufacturer[1]=%lx Product[1]=%lx\n",Manufac[1],Product[1]));
 #endif
@@ -915,3 +924,3 @@
 
-Matchup(USHORT slot,UCHAR *buttons,UCHAR *rtype)
+int Matchup(USHORT slot,UCHAR *buttons,UCHAR *rtype)
 {
@@ -938,3 +947,3 @@
 #ifdef DEBUG_DETECT
-                        prf("Matchup: NewType[%x]=%X\r\n",slot,NewType[slot]);
+                        dprintf(("Matchup: NewType[%x]=%lx\n",slot,NewType[slot]));
 #endif
@@ -976,3 +985,3 @@
 #ifdef DEBUG_DETECT
-                        prf("PickPDI: %x Not a built-in device type\r\n",j);
+                        dprintf(("PickPDI: %x Not a built-in device type\n",j));
 #endif
@@ -981,3 +990,3 @@
 #ifdef DEBUG_DETECT
-                                prf("PickPDI: %x Attempting to fallback\r\n");
+                                dprintf(("PickPDI: %x Attempting to fallback\n"));
 #endif
@@ -987,3 +996,3 @@
 #ifdef DEBUG_DETECT
-                                        prf("PickPDI: %x Supporting as generic PS/2 mouse\r\n");
+                                        dprintf(("PickPDI: %x Supporting as generic PS/2 mouse\n"));
 #endif
@@ -996,3 +1005,3 @@
 #ifdef DEBUG_DETECT
-                        prf("PickPDI: %x is a built-in device type\r\n",j);
+                        dprintf(("PickPDI: %x is a built-in device type\n",j));
 #endif
@@ -1002,3 +1011,3 @@
 #ifdef DEBUG_DETECT
-        prf("PickPDIType found %x devices.\r\n",found);
+        dprintf(("PickPDIType found %x devices.\n",found));
 #endif
diff -rwU1 VBoxMouse.org/pdiproto.h VBoxMouse/pdiproto.h
--- VBoxMouse.org/pdiproto.h	2000-09-06 10:53:52.000000000 +0200
+++ VBoxMouse/pdiproto.h	2026-05-06 13:35:30.573304500 +0200
@@ -64,2 +64,4 @@
 
+USHORT LgtchMagicKnock(void);
+
 #endif  // INCL_PDIPROTO
diff -rwU1 VBoxMouse.org/pdistick.c VBoxMouse/pdistick.c
--- VBoxMouse.org/pdistick.c	2000-09-06 10:56:52.000000000 +0200
+++ VBoxMouse/pdistick.c	2026-05-06 13:36:04.568321500 +0200
@@ -50,2 +50,3 @@
 #include "alltypes.h"
+#include "VBoxMouse.h"
 
@@ -326,3 +327,3 @@
                         #ifdef DEBUG
-                        prf("usValue=%xh\r\n", usValue);
+                        prf("usValue=%xh\n", usValue);
                         #endif
@@ -534,6 +535,6 @@
 USHORT
-LgtchMagicKnock()
+LgtchMagicKnock(void)
 {
 #ifdef DEBUG_L2
-        prf("LgtchMagicKnock\r\n");
+        dprintf(("LgtchMagicKnock\n"));
 #endif
@@ -557,3 +558,3 @@
 #ifdef DEBUG_L2
-        prf("MagicKnock\r\n");
+        dprintf(("MagicKnock\n"));
 #endif
@@ -580,3 +581,3 @@
 #ifdef DEBUG_L2
-        prf("MagicUnKnock\r\n");
+        dprintf(("MagicUnKnock\n"));
 #endif
@@ -611,3 +612,3 @@
 #ifdef DEBUG
-   prf("LgtchMouseInit()\r\n");
+        dprintf(("LgtchMouseInit()\n"));
 #endif
@@ -715,3 +716,3 @@
 #ifdef DEBUG_PACKET_ALL
-                prf("Packet %x: %x %x %x\r\n", plusType,gMEvent[0],gMEvent[1],gMEvent[2]);
+                dprintf(("Packet %x: %x %x %x\n", plusType,gMEvent[0],gMEvent[1],gMEvent[2]));
 #endif
@@ -722,3 +723,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring malformed packet\r\n");
+                                dprintf(("  Ignoring malformed packet\n"));
 #endif
@@ -728,3 +729,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring Device Type packet\r\n");
+                                dprintf(("  Ignoring Device Type packet\n"));
 #endif
@@ -734,3 +735,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Processing roller packet\r\n");
+                                dprintf(("  Processing roller packet\n"));
 #endif
@@ -750,3 +751,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Processing stickmouse packet\r\n");
+                                dprintf(("  Processing stickmouse packet\n"));
 #endif
@@ -758,3 +759,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring Touchpad extra info\r\n");
+                                dprintf(("  Ignoring Touchpad extra info\n"));
 #endif
@@ -764,3 +765,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring Wireless info\r\n");
+                                dprintf(("  Ignoring Wireless info\n"));
 #endif
@@ -770,3 +771,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring Wireless info\r\n");
+                                dprintf(("  Ignoring Wireless info\n"));
 #endif
@@ -776,3 +777,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring Wireless info\r\n");
+                                dprintf(("  Ignoring Wireless info\n"));
 #endif
@@ -782,3 +783,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring reserved packet\r\n");
+                                dprintf(("  Ignoring reserved packet\n"));
 #endif
@@ -788,3 +789,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring Wireless info\r\n");
+                                dprintf(("  Ignoring Wireless info\n"));
 #endif
@@ -794,3 +795,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring Wireless info\r\n");
+                                dprintf(("  Ignoring Wireless info\n"));
 #endif
@@ -800,3 +801,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring reserved packet\r\n");
+                                dprintf(("  Ignoring reserved packet\n"));
 #endif
@@ -806,3 +807,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring reserved packet\r\n");
+                                dprintf(("  Ignoring reserved packet\n"));
 #endif
@@ -812,3 +813,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring remote control LSB\r\n");
+                                dprintf(("  Ignoring remote control LSB\n"));
 #endif
@@ -818,3 +819,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring remote control MSB\r\n");
+                                dprintf(("  Ignoring remote control MSB\n"));
 #endif
@@ -824,3 +825,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring reserved packet\r\n");
+                                dprintf(("  Ignoring reserved packet\n"));
 #endif
@@ -830,3 +831,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring reserved packet\r\n");
+                                dprintf(("  Ignoring reserved packet\n"));
 #endif
@@ -841,3 +842,3 @@
 #ifdef DEBUG_PACKET
-                                prf("  Ignoring misdetected packet\r\n");
+                                dprintf(("  Ignoring misdetected packet\n"));
 #endif
@@ -862,3 +863,3 @@
 
-        ษอออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออป
+        +-----------------------------------------------------------------------+
         *        Bit 7   Bit 6   Bit 5   Bit 4   Bit 3   Bit 2   Bit 1   Bit 0  *
@@ -871,3 +872,3 @@
         *                                                                       *
-        ศอออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออผ
+        +-----------------------------------------------------------------------+
 
@@ -876,3 +877,3 @@
 
-        ษอออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออป
+        +-----------------------------------------------------------------------+
         *        Bit 7   Bit 6   Bit 5   Bit 4   Bit 3   Bit 2   Bit 1   Bit 0  *
@@ -885,3 +886,3 @@
         *                                                                       *
-        ศอออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออออผ
+        +-----------------------------------------------------------------------+
 
@@ -1021,3 +1022,3 @@
                         #ifdef DEBUG
-                        prf("newkeys = %xh\r\n", newkeys);
+                        prf("newkeys = %xh\n", newkeys);
                         #endif
@@ -1056,5 +1057,3 @@
 
-                #ifdef DEBUG
-                prf("y=%xh x=%xh event=%xh\r\n", y, x, event);
-                #endif
+                dprintf(("y=%xh x=%xh event=%xh\n", y, x, event));
 
@@ -1069,5 +1068,3 @@
 {
-        #ifdef DEBUG
-        prf("Primax init rate=%d resol=%d scale=%d\r\n", rate, resol, scale);
-        #endif
+        dprintf(("Primax init rate=%d resol=%d scale=%d\n", rate, resol, scale));
 
@@ -1079,5 +1076,3 @@
         error:
-        #ifdef DEBUG
-        prf("PS2_PrimaxInit: GeneriInit failed!\r\n");
-        #endif
+        dprintf(("PS2_PrimaxInit: GeneriInit failed!\n"));
         return (1);
diff -rwU1 VBoxMouse.org/pditrack.c VBoxMouse/pditrack.c
--- VBoxMouse.org/pditrack.c	2000-09-06 10:57:22.000000000 +0200
+++ VBoxMouse/pditrack.c	2020-03-04 19:29:25.680758400 +0100
@@ -19,2 +19,3 @@
 #include "alltypes.h"
+#include "VBoxMouse.h"
 
@@ -24,5 +25,3 @@
 {
-#ifdef DEBUG
-    prf("Trackpoint Transparent\r\n");
-#endif
+    dprintf(("Trackpoint Transparent\n"));
     if (SendToMouse(MOUCMD_TRACKPOINT))          goto exit;
@@ -40,5 +39,3 @@
 {
-#ifdef DEBUG
-    prf("Trackpoint Opaque\r\n");
-#endif
+    dprintf(("Trackpoint Opaque\n"));
     if (SendToMouse(MOUCMD_TRACKPOINT))          goto exit;
Only in VBoxMouse/: README.md
diff -rwU1 VBoxMouse.org/strat.asm VBoxMouse/strat.asm
--- VBoxMouse.org/strat.asm	2000-09-06 11:04:16.000000000 +0200
+++ VBoxMouse/strat.asm	2026-05-06 13:36:58.174201700 +0200
@@ -111,2 +111,5 @@
        extrn  Emi_Config                            : near      ;emi
+ifdef VBOXMOUSE
+       extrn  VBoxInitComplete                      : far       ; util1.asm
+endif
 
@@ -252,3 +255,3 @@
        LocalVar   TempES, WORD                ; Setup local stack frame
-       Localvar   TempBX, WORD
+       LocalVar   TempBX, WORD
 
@@ -296,3 +299,3 @@
        .if <<word ptr Device_Help+2> ne 0> NEAR AND
-       .if <Deinstall eq OFF> near
+       .if <DeInstall eq OFF> near
 
@@ -316,3 +319,3 @@
              mov  dl, DevHlp_SemRequest      ; Specify function number
-             call Device_help                ; Invoke SemRequest function
+             call Device_Help                ; Invoke SemRequest function
           .else                              ; if a close request then
@@ -459,3 +462,3 @@
           mov  dl, DevHlp_DevDone             ; Specify function number
-          call Device_help                    ; Invoke DevDone Function
+          call Device_Help                    ; Invoke DevDone Function
 
@@ -605,3 +608,3 @@
        .if <fs:Detach_Proc eq OFF> near       ; If Not Detached Process
-          test [si].D_Status, SQ_Mode         ; Check SQ Mode Flag Bit
+          test [si].D_Status, SQ_MODE         ; Check SQ Mode Flag Bit
           .if <z> near                        ; If SQ Mode InActive
@@ -614,3 +617,3 @@
 
-             test [si].D_Status, Busy_Mask    ; Check Que Busy Flag Bit
+             test [si].D_Status, BUSY_MASK    ; Check Que Busy Flag Bit
              .if <z> near                     ; If Event Que not busy with I/O
@@ -618,3 +621,3 @@
                 .if c                         ; If MonFlush failed then
-                   and  [si].D_Status, NOT Q_Flush  ; Reset Flush Flag Bit
+                   and  [si].D_Status, NOT Q_FLUSH  ; Reset Flush Flag Bit
                    GenFail                    ; Set Request Block Error code
@@ -622,3 +625,3 @@
 
-                and  [si].D_Status, NOT Busy_Mask  ; Reset Que Busy Flag
+                and  [si].D_Status, NOT BUSY_MASK     ; Reset Que Busy Flag
 
@@ -777,3 +780,3 @@
           xor  ch, ch                   ; Clear Settable Flags
-          and  cl, NOT Block_Mask       ; clr blocked bit
+          and  cl, NOT BLOCK_MASK       ; clr blocked bit
           mov  [si].D_Status, cx        ; Restore Dev Status Flags
@@ -800,3 +803,3 @@
              mov  [si].D_Status, ax        ; Init Device Status Flags
-             or   [si].D_Status, USS_Mode  ; Set Bad Display Mode Flag Bit
+             or   [si].D_Status, USS_MODE  ; Set Bad Display Mode Flag Bit
              mov  [si].Ptr_Row_Pos, ax     ; Set Ptr at upper left
@@ -808,3 +811,3 @@
 
-             test [si].MType, Graphics     ; Check current Display Mode
+             test [si].Mtype, Graphics     ; Check current Display Mode
              .if  nz                       ; If in Graphics Mode then
@@ -977,3 +980,3 @@
 
-         test [SI].D_Status, PtrDraw+USS_Mode ; Try PtrDD/Disp Mode Flags
+         test [SI].D_Status, PTRDRAW+USS_MODE ; Try PtrDD/Disp Mode Flags
          .if <z> AND                          ; If Ptr DD Used & Mode AOK
@@ -1365,2 +1368,5 @@
        call RegisterForAPM
+ifdef VBOXMOUSE
+       call VBoxInitComplete
+endif
 
diff -rwU1 VBoxMouse.org/usbstuff.c VBoxMouse/usbstuff.c
--- VBoxMouse.org/usbstuff.c	2000-09-06 10:57:32.000000000 +0200
+++ VBoxMouse/usbstuff.c	2020-03-04 19:29:25.607685300 +0100
@@ -50,2 +50,3 @@
 #include "alltypes.h"
+#include "VBoxMouse.h"
 
@@ -97,3 +98,3 @@
                               "push ax"\
-                              "call far _usbhandler"\
+                              "call far ptr _usbhandler"\
                               "add sp, 6"\
@@ -117,3 +118,2 @@
                  } else {
-                          if ( dwheel >= 128 ) {
                                         KbdEvent(2, (PSZ)&scancodes[1]);
@@ -121,3 +121,2 @@
                  }
-        }
 
@@ -135,2 +134,56 @@
 
+//@@@AH
+
+// assembler function, tell compiler to not decorate function name
+void _cdecl Process_Absolute();
+#pragma aux Process_Absolute "*";
+
+void SendUSBMouseMoveAbsolute(RP_GENIOCTL __far *pRP)
+{
+        MouseDataAbsolute FAR *MDAbs;
+        UCHAR dwheel;
+
+        MDAbs = (MouseDataAbsolute FAR *)pRP->ParmPacket;
+        dwheel = MDAbs->wheel;
+
+        // put data into global structure for absolute positions
+        Int_Packet.X_Pos = MDAbs->positionX;
+        Int_Packet.Y_Pos = MDAbs->positionY;
+        Int_Packet.X_Size = MDAbs->resolutionX;
+        Int_Packet.Y_Size = MDAbs->resolutionY;
+        Int_Packet.Event = MDAbs->deviceStatus;
+
+// @@@AH this trashes the stack, don't know why :-/
+#if 0
+        PUSH_STACK();
+
+        if ( dwheel ) {
+                 if ( dwheel < 128 ) {
+                          KbdEvent(2, (PSZ)&scancodes[0]) ;
+                 } else {
+                          if ( dwheel >= 128 ) {
+                                        KbdEvent(2, (PSZ)&scancodes[1]);
+                          }
+                 }
+        }
+
+        POP_STACK();
+#endif
+
+        // save our stack and call ASM function to process the packet
+        _asm
+        {
+           push ss
+           push bp
+           push sp
+           call Process_Absolute
+           pop sp
+           pop bp
+           pop ss
+        }
+
+        return;
+}
+//@@@AH
+
 void __far _cdecl IDCRoutine( RP_GENIOCTL __far *pRP)
@@ -151,5 +204,11 @@
 
+//@@@AH
+                        case USB_IDC_FUNCTION_PRCIRQABS:
+                                SendUSBMouseMoveAbsolute(pRP);
+                                break;
+//@@@AH
+
                         case USB_IDC_FUNCTION_REGISTER:
 #ifdef LDEBUG
-                                prf("MOUSE :  IDC call USB_IDC_FUNCTION_REGISTER\r\n");
+                                dprintf(("MOUSE :  IDC call USB_IDC_FUNCTION_REGISTER\n");
 #endif
@@ -165,3 +224,3 @@
 #ifdef LDEBUG
-                                prf("MOUSE :  IDC call USB_IDC_FUNCTION_SINGLE_DEVICE\r\n");
+                                dprintf(("MOUSE :  IDC call USB_IDC_FUNCTION_SINGLE_DEVICE\n");
 #endif
diff -rwU1 VBoxMouse.org/usbstuff.h VBoxMouse/usbstuff.h
--- VBoxMouse.org/usbstuff.h	2000-09-06 10:54:28.000000000 +0200
+++ VBoxMouse/usbstuff.h	2020-03-04 19:29:25.614692100 +0100
@@ -107,2 +107,6 @@
 
+//@@@AH
+   #define USB_IDC_FUNCTION_PRCIRQABS 0x60
+//@@@AH
+
 
@@ -123,3 +127,3 @@
 void call_GIDC(void);
-   #pragma aux call_GIDC = "call far GIDC_Entry";
+   #pragma aux call_GIDC = "call far ptr GIDC_Entry";
 
@@ -172,4 +176,7 @@
    #pragma alloc_text( RMCode, SendUSBMouseMove)
+//@@@AH
+   #pragma alloc_text( RMCode, SendUSBMouseMoveAbsolute)
+//@@@AH
 
-extern _far _usbhandler();
+extern void _far _usbhandler();
 extern void StickPacket();
@@ -190,2 +197,14 @@
 
+//@@@AH
+typedef struct _MouseDataAbsolute
+{
+   UCHAR deviceStatus; /* mouse status value */
+   USHORT positionX; /* absolute X position (device) */
+   USHORT positionY; /* absolute Y position (device) */
+   SHORT wheel; /* relative wheel coordinate */
+   USHORT resolutionX; /* X resolution of device */
+   USHORT resolutionY; /* Y resolution of device */
+} MouseDataAbsolute;
+//@@@AH
+
 
diff -rwU1 VBoxMouse.org/util1.asm VBoxMouse/util1.asm
--- VBoxMouse.org/util1.asm	2000-09-07 08:39:46.000000000 +0200
+++ VBoxMouse/util1.asm	2020-03-04 19:29:25.673751400 +0100
@@ -99,2 +99,7 @@
 
+ifdef VBOXMOUSE
+        extrn   _KbdDI_Handle   : word
+        extrn   _KbdDI_Entry    : dword
+endif
+
 CSEG    SEGMENT   WORD  PUBLIC  USE16 'CODE'
@@ -653,2 +658,87 @@
 
+ifdef VBOXMOUSE
+;;
+; Just a far call thunk - 16-bit is sooo fun.
+public VBoxInitComplete
+VBoxInitComplete PROC FAR
+       extrn    _VBoxAttachToVBoxGuest: near
+       ;
+       ; I'm going a bit overboard saving/restoring here...
+       ;
+       pushad
+       push     ds
+       push     es
+       push     fs
+       push     gs
+       mov      ax, seg APMDD
+       mov      es, ax
+       mov      ds, ax
+
+       call     _VBoxAttachToVBoxGuest
+
+       pop      gs
+       pop      fs
+       pop      es
+       pop      ds
+       popad
+       retf
+VBoxInitComplete ENDP
+
+;; USHORT __cdecl Call_KbdDI(int func,char __far  *buffer)
+;
+; Seems to be a little bit of confusion as to whether
+; who is cleaning up the stack, or at least if Watcom is
+; involed. So, I just play safe here and don't clean
+; up anything.
+;
+public _Call_KbdDI
+extrn _kbd_idc_entry:far
+_Call_KbdDI PROC NEAR
+        push    ebp
+        mov     ebp, esp
+        push    esi
+        push    edi
+        push    ds
+        push    es
+
+        ;
+        ; Connect?
+        ;
+        mov     ax, [_KbdDI_Handle]
+        cmp     ax, 0ffffh
+        jne     ckdi_1
+
+        push    0
+        lea     ax, [_kbd_idc_entry]
+        push    ax
+        push    ds
+        push    cs
+        call dword ptr [_KbdDI_Entry]
+        mov     [_KbdDI_Handle], ax
+
+        ;
+        ; Push the request.
+        ;
+ckdi_1:
+        mov     si, [bp + 6]            ; func
+        push    si
+;        mov     edi, [bp + 8]           ; buffer
+;        push    edi
+        les     di, [bp + 8]            ; buffer
+        push    di
+        push    es
+        push    ax                      ; kdbDI_Handle
+        call dword ptr ds:[_KbdDI_Entry]
+
+        lea     sp, [bp - 0ch]
+        pop     ds
+        pop     es
+        pop     esi
+        pop     edi
+        pop     ebp
+        ret
+_Call_KbdDI ENDP
+
+endif ; VBOXMOUSE
+
 
diff -rwU1 VBoxMouse.org/util2.asm VBoxMouse/util2.asm
--- VBoxMouse.org/util2.asm	2000-09-06 11:04:56.000000000 +0200
+++ VBoxMouse/util2.asm	2026-05-06 13:38:49.148983600 +0200
@@ -91,2 +91,6 @@
 
+IFDEF VBOXMOUSE
+       extrn  _VBoxUpdatePointer  : far
+ENDIF
+
        EXTRNFAR CollisionChk
@@ -1264,2 +1268,5 @@
                    call DrawPointer       ; GRADD
+IFDEF VBOXMOUSE
+                   call _VBoxUpdatePointer
+ENDIF
                 .endif
Only in VBoxMouse/: VBoxMouse.c
Only in VBoxMouse/: VBoxMouse.h
Only in VBoxMouse/: VBoxMouseDebug.c
diff -rwU1 VBoxMouse.org/vdm.asm VBoxMouse/vdm.asm
--- VBoxMouse.org/vdm.asm	2000-09-06 11:05:06.000000000 +0200
+++ VBoxMouse/vdm.asm	2026-05-06 13:39:21.086502500 +0200
@@ -77,2 +77,8 @@
 
+IFDEF VBOXMOUSE
+IFDEF DEBUG
+       extrn _VBoxScreenSizeChange  : far
+ENDIF
+ENDIF
+
 DSEG   SEGMENT   WORD  PUBLIC  USE16 'DATA'
@@ -539,2 +545,16 @@
 
+IFDEF VBOXMOUSE
+IFDEF DEBUG
+     push ax
+     push EmiSSI.Ssi_GRow_Res
+     push EmiSSI.Ssi_GCol_Res
+     push EmiSSI.Ssi_TRow_Res
+     push EmiSSI.Ssi_TCol_Res
+     movzx ax, EmiSSI.Ssi_Mtype
+     push  ax
+     call _VBoxScreenSizeChange
+     add  sp, 10
+     pop   ax
+ENDIF
+ENDIF
   ret
