--- a/makeself.sh	2025-12-10 11:14:05.000000000 +0100
+++ b/makeself.sh	2026-03-24 18:18:56.572066587 +0100
@@ -107,6 +107,7 @@
     echo "    --nooverwrite      : Do not extract the archive if the specified target directory exists"
     echo "    --tar-format opt   : Specify a tar archive format (default is ustar)"
     echo "    --tar-extra opt    : Append more options to the tar command line"
+    echo "    --tar-no-format    : The tar/gtar command does not support the --format option"
     echo "    --untar-extra opt  : Append more options to the during the extraction of the tar archive"
     echo "    --nomd5            : Don't calculate an MD5 for archive"
     echo "    --nocrc            : Don't calculate a CRC for archive"
@@ -186,6 +187,7 @@
 OWNERSHIP=n
 SIGN=n
 GPG_PASSPHRASE=""
+TAR_FORMAT_SUPPORTED=y
 
 # LSM file stuff
 LSM_CMD="echo No LSM. >> \"\$archname\""
@@ -315,6 +317,10 @@
 	    TAR_EXTRA="$2"
         shift 2 || { MS_Usage; exit 1; }
     ;;
+    --tar-no-format)
+	TAR_FORMAT_SUPPORTED=n
+    shift
+    ;;
     --untar-extra)
         UNTAR_EXTRA="$2"
         shift 2 || { MS_Usage; exit 1; }
@@ -655,6 +661,13 @@
 test -x "$TAR" || TAR=`exec <&- 2>&-; which bsdtar || command -v bsdtar || type bsdtar`
 test -x "$TAR" || TAR=tar
 
+if test "$TAR_FORMAT_SUPPORTED" = "y"; then
+	TAR_FORMAT_ARGS="--format $TAR_FORMAT"
+else
+	# Default behavior in older versions of makeself
+	TAR_FORMAT_ARGS=""
+fi
+
 tmparch="${TMPDIR:-/tmp}/mkself$$.tar"
 (
     if test "$APPEND" = "y"; then
@@ -684,7 +697,7 @@
         \) -print \
         | LC_ALL=C sort \
         | sed 's/./\\&/g' \
-        | xargs $TAR $TAR_EXTRA --format $TAR_FORMAT -$TAR_ARGS "$tmparch"
+        | xargs $TAR $TAR_EXTRA $TAR_FORMAT_ARGS -$TAR_ARGS "$tmparch"
 ) || {
     echo "ERROR: failed to create temporary archive: $tmparch"
     rm -f "$tmparch" "$tmpfile"
